concordium-contracts-common-derive/src/derive.rs

```diff
@@ -1302,48 +1641,106 @@ pub fn impl_deserial_with_state(ast: &syn::DeriveInput) -> syn::Result<TokenStre
-                // down.
-                let (tag_lit, tag_span) = variant_attributes.tag.map_or_else(
-                    || {
-                        (
-                            syn::LitInt::new(i.to_string().as_str(), variant.span()),
-                            variant.ident.span(),
-                        )
-                    },
-                    |tag| (tag.value, tag.span),
-                );
 
                 let variant_ident = &variant.ident;
-                matches_tokens.extend(quote! {
-                    #tag_lit => {
-                        #field_tokens
-                        Ok(#data_name::#variant_ident #pattern)
-                    },
-                });
+                if variant_attributes.forwards.is_empty() {
+                    let field_tokens: proc_macro2::TokenStream = field_names
+                        .iter()
+                        .zip(variant.fields.iter())
+                        .map(|(name, field)| {
+                            impl_deserial_with_state_field(
+                                field,
+                                &state_ident,
+                                name,
+                                &source,
+                                state_parameter,
+                            )
+                        })
+                        .collect::<syn::Result<proc_macro2::TokenStream>>()?;
 
-                tag_checker.add_and_check(tag_lit, tag_span, &variant.ident)?;
+                    // Get the literal for the tag either from a 'tag' attribute of the index of the
+                    // variant. To avoid cloning, the collision checking of tags are done further
+                    // down.
+                    let (tag_lit, tag_span) = variant_attributes.tag.map_or_else(
+                        || {
+                            (
+                                syn::LitInt::new(i.to_string().as_str(), variant.span()),
+                                variant_ident.span(),
+                            )
+                        },
+                        |tag| (tag.value, tag.span),
+                    );
+
+                    matches_tokens.extend(quote! {
+                        #tag_lit => {
+                            #field_tokens
+                            Ok(#data_name::#variant_ident #pattern)
+                        },
+                    });
+
+                    tag_checker.add_and_check(tag_lit, tag_span, &variant.ident)?;
+                } else {
+                    some_variant_is_forwarded = true;
+                    check!(
+                        variant.fields.len() == 1,
+                        variant.span(),
+                        "Only enum variants containing a single field can be used with the \
+                         'forward' attribute."
+                    );
+                    let chained_source = format_ident!("___chained_source");
```

Comment:
Unfortunately `def_site` is part of the unstable API of proc_macro2, but I went over this again and I don't see anyway for variables to collide with this identifier.

Reply:
Unfortunately `def_site` is part of the unstable api of `def_site`. I looked it over again and the fields of the variant are introduced as variables which might collide with this name. It happens to still work, since this case will only have one field, so losing the old variable is not a problem... I'll see if I can make it less a coincidence...

Reply:
Ok, unfortunate as it is :)